src/hooks/Auth.ts

```diff
@@ -97,84 +119,107 @@ export const useAuth = ({
+          if (statusCode >= 400 && statusCode < 500) {
+            // HTTP client error -> token is probably expired
+            console.log('Removing expired refresh token');
+            setRefreshToken(undefined);
+            setToken(undefined);
+            setIdToken(undefined);
+            await storage.removeItem(AUTH_REFRESH_TOKEN_KEY);
+          }
+        }
+        onError(err, ErrorAction.REFRESH_TOKEN_REQUEST);
+      }
+    },
+    [configuration, options.clientId, options.redirectUrl, options.tokenRequest?.extras, setTokenResponse, onError],
+  );
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const performTokenRefresh = useCallback(singleEntry(_refreshAccessToken), [_refreshAccessToken]);
+
   // Auto login if refresh token exists.
   useEffect(() => {
     void (async () => {
       const savedRefreshToken = await storage.getItem(AUTH_REFRESH_TOKEN_KEY);
-      if (!savedRefreshToken) {
+      if (!savedRefreshToken || isAutoLoginDone) {
         setIsAutoLoginDone(true);
         return;
       }
       if (configuration) {
         try {
-          const response = await performRefreshTokenRequest(
-            configuration,
-            options.clientId,
-            options.redirectUrl,
-            savedRefreshToken,
-            options.tokenRequest?.extras,
-          );
-          setTokenResponse(response);
-        } catch (err) {
-          if (err instanceof AppAuthError) {
-            const statusCode = Number.parseInt(err.message, 10);
-            if (statusCode >= 400 && statusCode < 500) {
-              // HTTP client error -> token is probably expired
-              console.log('Removing expired refresh token');
-              await storage.removeItem(AUTH_REFRESH_TOKEN_KEY);
-              setIsAutoLoginDone(true);
-              return;
-            }
-          }
-          onError(err, ErrorAction.AUTO_LOGIN);
+          await performTokenRefresh(savedRefreshToken);
+        } finally {
           setIsAutoLoginDone(true);
         }
       }
     })();
-  }, [configuration, onError, options.clientId, options.redirectUrl, options.tokenRequest?.extras, setTokenResponse]);
+  }, [configuration, isAutoLoginDone, onError, performTokenRefresh, setTokenResponse]);
 
   // Refresh periodically.
   useEffect(() => {
+    if (!isLoggedIn || !refreshToken || options.disableTokenRefresh) {
+      return;
+    }
     const timeoutId = setTimeout(async () => {
-      if (!isLoggedIn || !refreshToken) {
-        return;
-      }
-
-      if (configuration) {
-        await performRefreshTokenRequest(configuration, options.clientId, options.redirectUrl, refreshToken, options.tokenRequest?.extras)
-          .then(setTokenResponse)
-          .catch(err => onError(err, ErrorAction.REFRESH_TOKEN_REQUEST));
+      if (configuration && refreshToken) {
+        await performTokenRefresh(refreshToken.token);
       }
-    }, refreshInterval);
+    }, refreshToken.expiresIn * 0.9);
```

Comment:
This 0.9 is used two times in the code, create a global constant or even better, optional option the user can pass.